### DISCUSSION TITLE
`dependency_overrides` requires unnecessary arguments with Protocol/ABC in Swagger

### DISCUSSION THREAD

**Discussion by GLEF1X** [fastapi/fastapi#8668] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from __future__ import annotations

import dataclasses
from typing import Protocol

import uvicorn
from fastapi import FastAPI, Depends, Response

app = FastAPI()


@dataclasses.dataclass()
class Foo:
    id: int


class BaseRepo(Protocol):
    def get_foo(self, some_id: int) -> Foo: ...


class ConcreteRepo:
    def get_foo(self, some_id: int) -> Foo:
        return Foo(id=some_id)


@app.get("/hello/world")
async def view(repo: BaseRepo = Depends(BaseRepo)):
    return Response(200)


app.dependency_overrides[BaseRepo] = lambda: ConcreteRepo()  # also without lambda works incorrect

if __name__ == '__main__':
    uvicorn.run(app)
```

### Description

* Open /docs route and see that API method requires *args and **kwargs for request (probably from metaclass of Protocol)

### Operating System

Linux

### FastAPI Version

0.70.0

### Python Version

3.9.7

**Comment by GLEF1X** [comment]
https://github.com/tiangolo/fastapi/pull/4119

**Comment by marple-git** [comment]
👌

**Comment by Tishka17** [comment]
Is there any plans to fix this? 

Dependency injection feature is totally unusable because of this:
* We cannot use classes with `__init__` as a dependency, because all init args are shown in swagger, though they are have no relation to API calls
* We cannot use Protocol as a dependency as well, because random args and kwargs appear in swgger. Though, protocols cannot have instances

There is no way to use types as a dependencies, only possible solution is to create stub functions which looks very dirty.

**Comment by maksimzayats** [comment]
This library probably solves this problem
https://github.com/MaximZayats/fastapi-better-di

**Comment by GLEF1X** [comment]
> This library probably solves this problem https://github.com/MaximZayats/fastapi-better-di

It's not a quite great solution of this issue due to what this lib actually does. Patching `FastAPI` is an implicit spike to overcome FastAPI DI's flaw. I think it would be better to apply my brief PR #4119

**Comment by SebastianKrings** [comment]
would this also allow to override dependencies taking arguments?
like this:

`@app.get("/hello/world")
async def view(repo: BaseRepo = Depends(BaseRepo("anyArg"))):
    return Response(200)

app.dependency_overrides[BaseRepo] = lambda: ConcreteRepo() #override even without giving an argument`

**Comment by tiangolo** [comment, marked as answer]
Hey there! I see that in your examples `BaseRepo` and `ConcreteRepo` don't have an `__init__` method, so they don't require other dependencies, don't require data from the request, don't integrate with OpenAPI or OAuth2. Why do you want to put them in dependencies instead of creating the instances in the code?

Using things in dependencies is normally for when there's a reason to use them, because otherwise, it's simpler and more explicit to call the code directly in your function.

I want to understand your use case to see why it would make sense to support this, to achieve what, and see if there's another way to do it.

**Comment by GLEF1X** [comment]
Hi, @tiangolo! Thank you for your response! Example with `BaseRepo` and `ConcreteRepo` is simplified and was made just to show that the problem exists.

I believe, usage of DI using `dependency_overrides` is more farsighted and for big applications extremely useful, because it follows DI conception and gains all advantages of this approach.
From [wiki](https://en.wikipedia.org/wiki/Dependency_injection):

> In [software engineering](https://en.wikipedia.org/wiki/Software_engineering), dependency injection is a [design pattern](https://en.wikipedia.org/wiki/Software_design_pattern) in which an [object](https://en.wikipedia.org/wiki/Object_(computer_science)) or [function](https://en.wikipedia.org/wiki/Subroutine) receives other objects or functions that it depends on. A form of [inversion of control](https://en.wikipedia.org/wiki/Inversion_of_control), dependency injection aims to [separate the concerns](https://en.wikipedia.org/wiki/Separation_of_concerns) of constructing objects and using them, leading to [loosely](https://en.wikipedia.org/wiki/Loose_coupling) [coupled](https://en.wikipedia.org/wiki/Coupling_(computer_programming)) programs.[[1]](https://en.wikipedia.org/wiki/Dependency_injection#cite_note-1)[[2]](https://en.wikipedia.org/wiki/Dependency_injection#cite_note-MarkSeeman2011P4-2)[[3]](https://en.wikipedia.org/wiki/Dependency_injection#cite_note-3) The pattern ensures that an object or function which wants to use a given [service](https://en.wikipedia.org/wiki/Service_(systems_architecture)) should not have to know how to construct those services. Instead, the receiving '[client](https://en.wikipedia.org/wiki/Client_(computing))' (object or function) is provided with its dependencies by external code (an 'injector'), which it is not aware of.[[4]](https://en.wikipedia.org/wiki/Dependency_injection#cite_note-HollywoodPrinciple.c2-4) Dependency injection helps by making implicit dependencies explicit and helps solve the following problems:[[5]](https://en.wikipedia.org/wiki/Dependency_injection#cite_note-5)
> 
> How can a [class](https://en.wikipedia.org/wiki/Class_(computer_programming)) be independent from the creation of the objects it depends on?
> * How can an application, and the objects it uses support different configurations?
> * How can the behavior of a piece of code be changed without editing it directly?
> * Fundamentally, dependency injection consists of [passing parameters to a method](https://en.wikipedia.org/wiki/Parameter_(computer_programming)).[[6]](https://en.wikipedia.org/wiki/Dependency_injection#cite_note-6)
> 
> Because the client does not build or [find the service](https://en.wikipedia.org/wiki/Service_locator_pattern) itself, it typically only needs to declare the [interfaces](https://en.wikipedia.org/wiki/Interface_(computing)) of the services it uses, rather than their concrete implementations. This makes it easier to change which services are actually used at runtime, especially in [statically-typed languages](https://en.wikipedia.org/wiki/Statically_typed_language) where changing the underlying objects would otherwise require re-compiling the source code. 

I want to emphasize this:

> **The pattern ensures that an object or function which wants to use a given [service](https://en.wikipedia.org/wiki/Service_(systems_architecture)) should not have to know how to construct those services.**

In practice approach with using `Depends` with `dependency_overrides` makes code much more cleaner. For instance, I need to pass a `UserRepository` instance to the request handler. In `main` function, where app is being created I do [something like this](https://github.com/GLEF1X/fastapi-project/blob/dev/src/utils/application_builder/api_installation.py#L104-L106)  - in this example i use a workaround(empty classes) for this issue, but with fix of this issue it would be just an interface(abc class or protocol or just a plain class). So in request handler now I can just use `Depends` to get instance of `UserRepository` - [code example](https://github.com/GLEF1X/fastapi-project/blob/dev/src/api/v1/endpoints/users.py#L21-L31). If I want to add caching or logging to `UserRepository` I won't touch the request handler's code. Instead of it, I will implement the [Decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern) and just change the code of `dependency_overrides`.

Maybe you will ask why it's better than something like the `get_db`. It's not bad approach, but it violates the DI philosophy.
```python
def get_db():
    session = session_maker()
    try:
        yield session
    finally:
        session.close()
 
@app.get("/")
async def handler(db: Session = Depends(get_db)):
    return {"message": "Hello World"}
```

In this example, the handler knows about the factory of dependency and strongly tied to it(the module with handler must import it).

**Comment by JarroVGIT** [comment]
> It's not bad approach, but it violates the DI philosophy.

But why is that? To re-quote what you wanted to emphasise:
> The pattern ensures that an object or function which wants to use a given [service](https://en.wikipedia.org/wiki/Service_(systems_architecture)) should not have to know how to construct those services.

Having a signature of `async def handler(db: Session = Depends(get_db)):` complies with that statement, no? The `handler` function here still has no notion of the construction of a Session object, only how to get an instance of it (so yes, it knows the factory). The argument `the handler now has a strongly tied because it needs to  import it now` is not very strong imho, because in the alternative it would need to import `BaseRepo` instead. 

I can see this making sense when you, on runtime, want to determine what implementation of `BaseRepo` must be passed to the path operation function, but that is also not possible with the proposed solution (as it would override all cases of `BaseRepo`). 

It is an interesting idea though!

**Comment by GLEF1X** [comment]
Hello, I've found related issue #617 with the same problem but with a more complicated solution. With `dependency_overrides`  the problem of this issue can be solved easily.

For instance this:
```python
from fastapi import FastAPI, Depends
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker, Session

app = FastAPI()
engine = create_engine("sqlite://")
session_maker = sessionmaker()


def get_db_session():
    session = session_maker()
    try:
        yield session
    finally:
        session.close()


@app.get("/")
async def root(session: Session = Depends(get_db_session)):
    pass

```

Can be written this way with `dependency_overrides`:
```python
import uvicorn
from fastapi import FastAPI, Depends
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker, Session


def provide_session(session_maker: sessionmaker):
    def inner():
        session = session_maker()
        try:
            yield session
        finally:
            session.close()

    return inner


def create_app():
    app = FastAPI()

    engine = create_engine("sqlite://", echo=True, echo_pool=True)
    session_maker = sessionmaker(engine)
    app.dependency_overrides[Session] = provide_session(session_maker)
    return app


app = create_app()


@app.get("/")
async def root(session: Session = Depends(Session)):
    print(session)


uvicorn.run(app)

```

No global variables - fewer problems

Another related issue: #4183

**Comment by jbmusso** [comment]
I'm having the same issue as @GLEF1X and I can confirm the fix in https://github.com/tiangolo/fastapi/pull/4119 solves this issue.

The above solution is how I supposed FastAPI would work:
* `args` and `kwargs` should not appear in Swagger when using Protocol
* I'd rather specify interfaces as `Protocol` in route handler rather than concrete implementations (which needs to be imported at the module-level, only to be overridden later in tests)

@GLEF1X 's solution is rather elegant and I believe is similar to several dependency injection approaches in other languages.

Roughly, I'm following a clean architecture / hexagonal architecture approach and only rely on interfaces (Protocol) for specifying dependencies.

edit: sorry, I hit shift+enter too quickly

**Comment by sotirr** [comment]
Is there any plans to fix this?

I want to have dedependencies that depends on interface(Protocol) not on concrete realisation.

**Comment by MoBoo** [comment]
Is there an update to this?

**Comment by Tishka17** [comment]
As there are no updates I can suggest my workaround: https://github.com/Tishka17/fastapi-template/blob/master/src/app/api/depends_stub.py

In the view function should should wrap your dependency with Stub so like this
```python
@users_router.get("/alternative")
def add_users_alternative(
        new_user: Annotated[Session, Depends(Stub(Session))],
) -> SomeResult:
    ...
```

You can also use Stub with real classes, not protocols: it makes fastapi ignore all init params
And then you can provide it as usual:
```python
def new_session(session_maker: sessionmaker) -> Iterable[Session]:
    with session_maker() as session:
        yield session

app.dependency_overrides[Session] = partial(new_session, session_maker)
```

**Comment by dolfinus** [comment]
Here #10898 is a proposal for adding new kind of annotation to FastAPI:
* `obj: Class = Depends()` - same behavior as before with passing mandatory arguments to OpenAPI schema.
* `obj: Class = Service()` - do not add mandatory fields to OpenAPI schema.

**Comment by purepani** [comment]
I would also appreciate any solution to this problem. Just wanted to add an example of a use case.
I have a protocol that I'm intending to use as an interface for anyone to implement, so that they can use the library I'm writing with any software as needed. As in, they implement the protocol, and they are able to use the library for another function. 

I'd definitely appreciate any continued discussion on this, as this would make my use case, and likely many other use cases, much simpler.

**Reply by Tishka17** [reply]
@purepani I am currently developing a standalone IoC-container which can be used not only with fastapi, but also with other frameworks.  From my perspective, it is a good idea to keep using `fastapi.Depends` for request decomposition and use another container injection of other objects. You can have a look here: https://github.com/reagento/dishka/blob/develop/examples/fastapi_app.py

**Comment by linde12** [comment]
4+ years without proper DI :-) Is there any blocker/opinion preventing this from being implemented? https://github.com/fastapi/fastapi/pull/10898/files looks promising but there has been no feedback since the PR was created over a year ago.

Having to use concrete types severely limits the usefulness of current DI solution.
